### PR TITLE
Proppant solver refactoring

### PR DESCRIPTION
**Pull request by francoishamon**
I completed a first pass at the refactoring of the proppant solvers. In the current PR, we can match again the old baselines of the two proppant integrated tests (no more numerical diffs).

I modified the flow/proppant solvers as follows:

In the single-phase flow solvers:
- [X] Removed entirely the `transTMultiplier` that previously passed to all the solvers
- [X] Flux assembled in the style "launch -> computeWeights -> computeJunction"
- [X] Permeability and porosity modified through calls to new constitutive models

In the proppant solvers:
- [X] Moved to a new folder `physicsSolvers/fluidFlow/proppantTransport`
- [X] Removed entirely the `transTMultiplier` and `poroMultiplier`
- [X] Flux and CellBasedFlux assembled in the style "launch -> computeWeights -> compute(Junction)" 
- [X] No more stencil computation hardcoded in the kernels

Please let me know if this is going in the right direction.

Requires rebaseline, done in https://github.com/GEOSX/integratedTests/pull/152

### CODE REVIEW

**Review by CusiniM — APPROVED** [review]
This looks good. Getting rid of that `transTMultiplier` should be considered a success by itself! I just wonder if we can move that `coefficientMultiplier` directly in the constitutive model and avoid having separate interfaces in the stencil.

File: src/coreComponents/finiteVolume/SurfaceElementStencil.hpp [hunk 1]
```diff
@@ -433,6 +470,101 @@ inline void SurfaceElementStencilWrapper::computeWeights( localIndex iconn,
...
+      localIndex const ei0  =  m_elementIndices[iconn][k[0]];
+
+      localIndex const er1  =  m_elementRegionIndices[iconn][k[1]];
+      localIndex const esr1 =  m_elementSubRegionIndices[iconn][k[1]];
+      localIndex const ei1  =  m_elementIndices[iconn][k[1]];
+
+      real64 const mult0 = ( LvArray::math::abs( LvArray::tensorOps::AiBi< 3 >( m_cellCenterToEdgeCenters[iconn][k[0]], gravityVector ) ) > 1e-10 )
+  ? coefficientMultiplier[er0][esr0][ei0][0][1] : coefficientMultiplier[er0][esr0][ei0][0][0];
+      real64 const mult1 = ( LvArray::math::abs( LvArray::tensorOps::AiBi< 3 >( m_cellCenterToEdgeCenters[iconn][k[1]], gravityVector ) ) > 1e-10 )
+  ? coefficientMultiplier[er1][esr1][ei1][0][1] : coefficientMultiplier[er1][esr1][ei1][0][0];
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/finiteVolume/SurfaceElementStencil.hpp:512]: Yes, I completely agree. The difficulty is that the multiplier is computed cell-wise, but is applied differently depending on the direction of the connection (vertical or horizontal). So we could try to explore two ways of solving this problem and reduce the number of `computeWeights` kernels:
- We could entirely remove the `coefficientMultiplier` and tell the permeability constitutive model named `ProppantPermeability` to update three values per cell (say, `k_x`, `k_y` (= `k_x`), and `k_z`). That would be relatively clean, but I would have to modify the existing `computeWeights` kernels to account for this new format of the permeability.
- Or we could keep the `coefficientMultiplier` but use it in a slightly different way (that would probably modify the baselines a little bit). In this alternative way, we would compute the transmissibility using the existing `computeWeights` function that you implemented, and then just after the call to `computeWeights`, we would scale the transmissibility by the harmonic average of the `coefficientMultiplier` (similar to [this](https://github.com/GEOSX/GEOSX/blob/d174b40b822501aec4fbadd674d39ab8bd87825b/src/coreComponents/physicsSolvers/multiphysics/SinglePhasePoromechanicsFluxKernels.cpp#L452), but with a harmonic average to make it as close to the original implementation as possible).

I can try quickly the second approach to see how much it changes the baselines.

File: src/coreComponents/finiteVolume/TwoPointFluxApproximation.cpp [hunk 2]
```diff
@@ -334,67 +334,70 @@ void TwoPointFluxApproximation::addToFractureStencil( MeshLevel & mesh,
     localIndex const numElems = fractureConnectorsToFaceElements.sizeOfArray( fci );
     // only do this if there are more than one element attached to the connector
     localIndex const edgeIndex = fractureConnectorsToEdges[fci];
-    if( numElems > 1 )  // if there is only one element it does not need to be in the stencil.
-    {
-      localIndex const connectorIndex = fractureStencil.size();
 
-      GEOSX_ERROR_IF( numElems > maxElems, "Max stencil size exceeded by fracture-fracture connector " << fci );
+    ///////////////////////// TODO FRANCOIS: THIS IS JUST TO MATCH PREVIOUS PROPPANT BASELINES ///////////////////
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/finiteVolume/TwoPointFluxApproximation.cpp:338]: Yes, I need to investigate this issue. Doing what you suggested was my initial plan, but I realized that if I filter out the connections with only one element, then the results of the `FlowProppantBedTransport` integrated test are completely different (nothing seems to move anymore). I wonder if the ProppantTransport kernels use these one-element connections to impose some kind of boundary condition. This is not clear to me yet, I will update once I fully understand.

**Comment by jhuang2601** [comment]
@francoishamon @rrsettgast Great job! I've reran the validation example with this commit and compared the modeling results with experimental data and previous work. As shown below, they match well with each other.
image

File: src/coreComponents/constitutive/permeability/ProppantPermeability.hpp [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    real64 const perm = 0.25 * ( oldHydraulicAperture*oldHydraulicAperture*oldHydraulicAperture +
+                                 oldHydraulicAperture*oldHydraulicAperture*newHydraulicAperture +
+                                 oldHydraulicAperture*newHydraulicAperture*newHydraulicAperture +
+                                 newHydraulicAperture*newHydraulicAperture*newHydraulicAperture ) / 12;
+
+    real64 const dPerm  = 0.25 * ( oldHydraulicAperture*oldHydraulicAperture +
+                                   2*oldHydraulicAperture*newHydraulicAperture +
+                                   3*newHydraulicAperture*newHydraulicAperture ) / 12;
+
+    real64 const squaredHydraulicAperture = newHydraulicAperture * newHydraulicAperture;
```
Comment (sytuannguyen) [inline src/coreComponents/constitutive/permeability/ProppantPermeability.hpp:64]:
```suggestion
    real64 const squaredHydraulicAperture = newHydraulicAperture * newHydraulicAperture;
    real64 const oldSquaredHydraulicAperture = oldHydraulicAperture * oldHydraulicAperture;
    
    real64 const perm = ( oldSquaredHydraulicAperture*oldHydraulicAperture +
                                 oldSquaredHydraulicAperture*newHydraulicAperture +
                                 oldHydraulicAperture*squaredHydraulicAperture +
                                 newHydraulicAperture*squaredHydraulicAperture ) / 48;

    real64 const dPerm  = ( oldSquaredHydraulicAperture +
                                   2*oldHydraulicAperture*newHydraulicAperture +
                                   3*squaredHydraulicAperture ) / 48;
```

File: src/coreComponents/finiteVolume/SurfaceElementStencil.hpp [hunk 4]
```diff
@@ -433,6 +470,101 @@ inline void SurfaceElementStencilWrapper::computeWeights( localIndex iconn,
...
+{
+  // TODO: this should become star-delta method
+  real64 sumOfTrans = 0.0;
+  for( localIndex k=0; k<numPointsInFlux( iconn ); ++k )
+  {
+    localIndex const er  =  m_elementRegionIndices[iconn][k];
+    localIndex const esr =  m_elementSubRegionIndices[iconn][k];
+    localIndex const ei  =  m_elementIndices[iconn][k];
+
+    real64 const mult = ( LvArray::math::abs( LvArray::tensorOps::AiBi< 3 >( m_cellCenterToEdgeCenters[iconn][k], gravityVector ) ) > 1e-10 )
```
Comment (sytuannguyen) [inline src/coreComponents/finiteVolume/SurfaceElementStencil.hpp:488]: This limit of 1e-10 is used several times in this class. Do you want to define a constant for it?

File: src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseProppantBase.cpp [hunk 5]
```diff
@@ -100,35 +126,68 @@ void SinglePhaseProppantBase::updateFluidModel( Group & dataGroup, localIndex co
...
+
+void SinglePhaseProppantBase::updatePorosityAndPermeability( SurfaceElementSubRegion & subRegion,
+                                                             localIndex const targetIndex ) const
+{
+  GEOSX_MARK_FUNCTION;
+
+  arrayView1d< real64 const > const proppantPackVolumeFraction = subRegion.getReference< array1d< real64 > >( ProppantTransport::viewKeyStruct::proppantPackVolumeFractionString() );
+
+  arrayView1d< real64 const > const newHydraulicAperture = subRegion.getReference< array1d< real64 > >( viewKeyStruct::effectiveApertureString() );
+  arrayView1d< real64 const > const oldHydraulicAperture = subRegion.getReference< array1d< real64 > >( viewKeyStruct:: viewKeyStruct::aperture0String() );
```
Comment (sytuannguyen) [inline src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseProppantBase.cpp:138]:
```suggestion
  arrayView1d< real64 const > const oldHydraulicAperture = subRegion.getReference< array1d< real64 > >( viewKeyStruct::aperture0String() );
```

File: src/coreComponents/physicsSolvers/multiphysics/SinglePhasePoromechanicsFluxKernels.hpp [hunk 6]
```diff
@@ -166,12 +160,53 @@ struct FaceElementFluxKernel
...
+   * @param[in] dDens_dPres The change in material density for each element
+   * @param[in] mob The fluid mobility in each element
+   * @param[in] dMob_dPres The derivative of mobility wrt pressure in each element
+   * @param[in] permeability
+   * @param[in] dPerm_dPres The derivative of permeability wrt pressure in each element
+   * @param[in] permeabilityMultiplier
+   * @param[in] gravityVector
+   * @param[out] localMatrix The linear system matrix
+   * @param[out] localRhs The linear system residual
+   */
```
Comment (sytuannguyen) [inline src/coreComponents/physicsSolvers/multiphysics/SinglePhasePoromechanicsFluxKernels.hpp:186]:
```suggestion
  /**
   * @brief launches the kernel to assemble the flux contributions to the linear system.
   * @tparam SurfaceElementStencilWrapper The type of the stencil that is being used.
   * @param[in] stencilWrapper The stencil wrapper object.
   * @param[in] dt The timestep for the integration step.
   * @param[in] rankOffset The rank offset
   * @param[in] pressureDofNumber The pressure dof number for each element
   * @param[in] ghostRank The ghost rank
   * @param[in] pres The pressures in each element
   * @param[in] dPres The pressure change in each element
   * @param[in] gravCoef The factor for gravity calculations (g*H)
   * @param[in] dens The material density in each element
   * @param[in] dDens_dPres The change in material density for each element
   * @param[in] mob The fluid mobility in each element
   * @param[in] dMob_dPres The derivative of mobility wrt pressure in each element
   * @param[in] permeability The permeability in each element
   * @param[in] dPerm_dPres The derivative of permeability wrt pressure in each element
   * @param[in] dPerm_dAper The derivative of permeability wrt aperture in each element
   * @param[in] permeabilityMultiplier The permeability multiplier
   * @param[in] gravityVector The gravity vector
   * @param[out] localMatrix The linear system matrix
   * @param[out] localRhs The linear system residual
   */
```

File: src/coreComponents/constitutive/solid/ProppantSolid.hpp [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+  /**
+   * @brief Get catalog name
+   * @return Catalog name string
+   */
+  virtual string getCatalogName() const override { return catalogName(); }
+
+
+  /**
+   * @brief Create a instantiation of the ProppantSolidUpdates class
```
Comment (sytuannguyen) [inline src/coreComponents/constitutive/solid/ProppantSolid.hpp:122]:
```suggestion
   * @brief Create an instantiation of the ProppantSolidUpdates class
```

File: src/coreComponents/constitutive/permeability/ProppantPermeability.hpp [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+    // horizontal multiplier
+    permeabilityMultiplier[0] = ( 1.0 - proppantPackVolumeFraction ) + 12.0 * proppantPackVolumeFraction * m_proppantPackPermeability / squaredHydraulicAperture;
+
+    // vertical multiplier
+    permeabilityMultiplier[1] = 1.0 - proppantPackVolumeFraction + proppantPackVolumeFraction * squaredHydraulicAperture / ( 12.0 * m_proppantPackPermeability );
+    permeabilityMultiplier[1] = 1.0 / permeabilityMultiplier[1];
```
Comment (sytuannguyen) [inline src/coreComponents/constitutive/permeability/ProppantPermeability.hpp:71]:
```suggestion
    permeabilityMultiplier[1] = 1.0 / ( 1.0 - proppantPackVolumeFraction + proppantPackVolumeFraction * squaredHydraulicAperture / ( 12.0 * m_proppantPackPermeability ) );
```

File: src/coreComponents/constitutive/permeability/ProppantPermeability.cpp [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+  // NOTE: enforcing 1 quadrature point
+  m_dPerm_dAperture.resize( 0, 1, 3 );
+  m_permeabilityMultiplier.resize( 0, 1, 3 );
+  PermeabilityBase::allocateConstitutiveData( parent, numConstitutivePointsPerParentIndex );
+}
+
+
+REGISTER_CATALOG_ENTRY( ConstitutiveBase, ProppantPermeability, string const &, Group * const )
+
+}
```
Comment (sytuannguyen) [inline src/coreComponents/constitutive/permeability/ProppantPermeability.cpp:79]:
```suggestion
} /* namespace constitutive */
```

File: src/coreComponents/constitutive/permeability/ProppantPermeability.hpp [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+  struct viewKeyStruct : public PermeabilityBase::viewKeyStruct
+  {
+    static constexpr char const * permeabilityMultiplierString() { return "permeabilityMultiplier"; }
+
+    static constexpr char const * maxProppantConcentrationString() { return "maxProppantConcentration"; }
+    static constexpr char const * proppantDiameterString() { return "proppantDiameter"; }
+
+    static constexpr char const * proppantPackPermeabilityString() { return "proppantPackPermeability"; }
+
+  } viewKeys;
```
Comment (sytuannguyen) [inline src/coreComponents/constitutive/permeability/ProppantPermeability.hpp:150]:
```suggestion
  };
```

File: src/coreComponents/constitutive/permeability/ProppantPermeability.hpp [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+
+  real64 m_proppantDiameter;
+
+  real64 m_maxProppantConcentration;
+
+  real64 m_proppantPackPermeability;
+
+};
+
+}/* namespace constitutive */
```
Comment (sytuannguyen) [inline src/coreComponents/constitutive/permeability/ProppantPermeability.hpp:170]:
```suggestion
} /* namespace constitutive */
```

File: src/coreComponents/constitutive/solid/ProppantSolid.cpp [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+template< typename PORO_TYPE,
+          typename PERM_TYPE >
+ProppantSolid< PORO_TYPE, PERM_TYPE >::~ProppantSolid() = default;
+
+typedef ProppantSolid< ProppantPorosity, ProppantPermeability > ProppantSolidModel;
+
+REGISTER_CATALOG_ENTRY( ConstitutiveBase, ProppantSolidModel, string const &, Group * const )
+
+}
```
Comment (sytuannguyen) [inline src/coreComponents/constitutive/solid/ProppantSolid.cpp:46]:
```suggestion
} /* namespace constitutive */
```

File: src/coreComponents/constitutive/solid/ProppantSolid.hpp [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+                                                         getPermModel() );
+  }
+private:
+  using CoupledSolid< NullModel, PORO_TYPE, PERM_TYPE >::getSolidModel;
+  using CoupledSolid< NullModel, PORO_TYPE, PERM_TYPE >::getPorosityModel;
+  using CoupledSolid< NullModel, PORO_TYPE, PERM_TYPE >::getPermModel;
+
+};
+
+}
```
Comment (sytuannguyen) [inline src/coreComponents/constitutive/solid/ProppantSolid.hpp:140]:
```suggestion
} /* namespace constitutve */
```

File: src/coreComponents/constitutive/solid/porosity/ProppantPorosity.cpp [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+void ProppantPorosity::postProcessInput()
+{
+  this->getWrapper< array1d< real64 > >( viewKeyStruct::referencePorosityString() ).
+    setApplyDefaultValue( m_defaultReferencePorosity );
+}
+
+REGISTER_CATALOG_ENTRY( ConstitutiveBase, ProppantPorosity, string const &, Group * const )
+}
```
Comment (sytuannguyen) [inline src/coreComponents/constitutive/solid/porosity/ProppantPorosity.cpp:60]:
```suggestion
} /* namespace constitutive */
```

**Review by sytuannguyen — COMMENTED** [review]
Just some minor suggestions.

File: src/coreComponents/constitutive/permeability/ProppantPermeability.hpp (hunk 3 above)
Comment (francoishamon) [reply src/coreComponents/constitutive/permeability/ProppantPermeability.hpp]: Done

Comment (francoishamon) [reply src/coreComponents/constitutive/permeability/ProppantPermeability.hpp]: Done

File: src/coreComponents/finiteVolume/SurfaceElementStencil.hpp (hunk 4 above)
Comment (francoishamon) [reply src/coreComponents/finiteVolume/SurfaceElementStencil.hpp]: Done

File: src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseProppantBase.cpp (hunk 5 above)
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseProppantBase.cpp]: Done

File: src/coreComponents/physicsSolvers/multiphysics/SinglePhasePoromechanicsFluxKernels.hpp (hunk 6 above)
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/multiphysics/SinglePhasePoromechanicsFluxKernels.hpp]: Done

File: src/coreComponents/constitutive/solid/ProppantSolid.hpp (hunk 7 above)
Comment (francoishamon) [reply src/coreComponents/constitutive/solid/ProppantSolid.hpp]: Done

File: src/coreComponents/constitutive/permeability/ProppantPermeability.hpp (hunk 8 above)
Comment (francoishamon) [reply src/coreComponents/constitutive/permeability/ProppantPermeability.hpp]: Done

File: src/coreComponents/constitutive/permeability/ProppantPermeability.cpp (hunk 9 above)
Comment (francoishamon) [reply src/coreComponents/constitutive/permeability/ProppantPermeability.cpp]: Done

File: src/coreComponents/constitutive/permeability/ProppantPermeability.hpp (hunk 10 above)
Comment (francoishamon) [reply src/coreComponents/constitutive/permeability/ProppantPermeability.hpp]: Done

File: src/coreComponents/constitutive/solid/ProppantSolid.cpp (hunk 12 above)
Comment (francoishamon) [reply src/coreComponents/constitutive/solid/ProppantSolid.cpp]: Done

File: src/coreComponents/constitutive/permeability/ProppantPermeability.hpp (hunk 11 above)
Comment (francoishamon) [reply src/coreComponents/constitutive/permeability/ProppantPermeability.hpp]: Done

File: src/coreComponents/constitutive/solid/ProppantSolid.hpp (hunk 13 above)
Comment (francoishamon) [reply src/coreComponents/constitutive/solid/ProppantSolid.hpp]: Done

File: src/coreComponents/constitutive/solid/porosity/ProppantPorosity.cpp (hunk 14 above)
Comment (francoishamon) [reply src/coreComponents/constitutive/solid/porosity/ProppantPorosity.cpp:60]: Done

**Review by sytuannguyen — APPROVED** [review]
